eventqueue/processor.go

```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eventqueue
+
+import (
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	kclock "k8s.io/utils/clock"
+)
+
+// ErrProcessorStopped is returned when the processor is not running.
+var ErrProcessorStopped = errors.New("processor is stopped")
+
+// Processor manages the queue of items and processes them at the correct time.
+type Processor[T queueable] struct {
+	executeFn          func(r T)
+	queue              *queue[T]
+	clock              kclock.Clock
+	queueLock          sync.Mutex
+	processorRunningCh chan struct{}
+	stopCh             chan struct{}
+	resetCh            chan struct{}
+	stopped            atomic.Bool
+}
+
+// NewProcessor returns a new Processor object.
+// executeFn is the callback invoked when the item is to be executed; this will be invoked in a background goroutine.
+func NewProcessor[T queueable](executeFn func(r T), clock kclock.Clock) *Processor[T] {
+	return &Processor[T]{
+		executeFn:          executeFn,
+		queue:              newQueue[T](),
+		processorRunningCh: make(chan struct{}, 1),
```

Comment:
This absolutely must have a capacity of 1 because of how we're using it.

https://go.dev/play/p/TK0y6Porva7

Then try to remove the capacity and see that it ends up in a deadlock